### PR TITLE
[Feature](bangc-ops): add new op three_nn_forward.

### PR DESCRIPTION
**Pull request by DanieeelLiu**
Thanks for your contribution and we appreciate it a lot. 

## 1. Motivation

Add new op three_nn_forward
## 2. Modification

modified:   bangc-ops/kernel_depends.toml
new file:   bangc-ops/kernels/three_nn_forward/three_nn_forward.cpp
new file:   bangc-ops/kernels/three_nn_forward/three_nn_forward_union1.mlu
modified:   bangc-ops/mlu_op.h
modified:   bangc-ops/mlu_op_kernel.h
new file:   bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/three_nn_forward/test_case/case0_.prototxt
new file:   bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/three_nn_forward/three_nn_forward.cpp
new file:   bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/three_nn_forward/three_nn_forward.h
new file:   docs/bangc-docs/design_docs/three_nn_forward/three_nn_forward.md
new file:   docs/bangc-docs/design_docs/three_nn_forward/threennsplit.png
new file:   docs/bangc-docs/design_docs/three_nn_forward/workflow.png
modified:   docs/bangc-docs/user_guide/9_operators/index.rst

## 3. Test Report

If you want to know how to do operator testing, you can see [GTest-User-Guide-zh](../docs/GTest-User-Guide-zh.md).

#### 3.1.1 Accuracy Acceptance Standard

Dynamic accuracy standard:
- [x] diff1,diff2, diff4 = [10, 10, 1] 

#### 3.1.2 Operator Scheme checklist

|     No.        |                 Details              |            Check Results             |
|----------------|--------------------------------------|--------------------------------------|
|        1       |Supported hardware                    |             MLU370<br>MLU590         |
|        2       |Job types                             |          block <br> U1 <br> U4       |
|        3       |Layouts                               |          NHWC 、NCHW、ARRAY etc      |
|        4       |Whether multi-dimensions are supported|                                      |
|        5       |Whether element zero is supported     |                                      |
|        6       |Data type(half/float)                 |           half / float etc           |
|        7       |Whether there is size limit           |                                      |

#### 3.1.3 New Feature Test

If you have checked the following items, please tick the relevant box.

- [x] Data type test
- [x] Multidimensional tensor test
- [x] Layout test
- [ ] Different size/integer remainder end segment/alignment misalignment test
- [x] Zero dimensional tensor test/zero element test
- [x] stability test
- [x] Multiple platform test
- [x] Gen_case module test
- [x] Nan/INF tests 
- [ ] Bug fix tests
- [x] For memory leak check details, see[GTest-User-Guide-zh](../docs/GTest-User-Guide-zh.md).
- [x] For code coverage check details, see: [GTest-User-Guide-zh](../docs/GTest-User-Guide-zh.md).
- [ ] For I/O calculation efficiency check details, see: [MLU-OPS Performance Acceptance Standard](../docs/MLU-OPS-Performance-Acceptance-Standard.md).

#### 3.1.4 Parameter Check

When a new operator is submitted, the test points are given and the test results are stated.

|                   Test Point                    | Acceptance Standard | Test Result (Error Message) |
| ----------------------------------------------- | --------------------| --------------------------- |
| Whether it conforms to the operator restriction |     Normal error    |                             |
| Whether illegal parameters are passed           |     Normal error    |                             |

### 3.2 Accuracy Test

For the cases used in the New Feature Test section, the features and the number of cases are recorded here. When multiple operations are tested, multiple tables are needed to include details of these operations.

Operation:

|Test Point           | Description                      | Quantity |  Comment |
|----------           |----------------------------------|----------|  --------|
|Data type test       |half/float/int8                   |          |          |
|Mult-tensor test     |Supports 1-8 dims                 |          |          |
|Layout test          |Supports NCHW/NHWC                |          |          |
|Zero element test    |Whether to support this test      |          |          |
|Stability test       |--gtest_repeat=NUM<br>--thread=NUM|          |          |
|Mult-platform test   |MLU370/MLU590                     |          |          |
|Nan/INF test         |Whether to support this test      |          |          |

### 3.3 Performance Test

See [MLU-OPS Performance Acceptance Standard](../docs/MLU-OPS-Performance-Acceptance-Standard.md) for details.

Platform：MLU370
release_test cases
[----------] Global test environment tear-down
[ SUMMARY  ] Total 50 cases of 1 op(s).
ALL PASSED.
[==========] 50 test cases from 1 test suite ran.
[  PASSED  ] 50 test cases.

release_temp cases
[----------] Global test environment tear-down
[ SUMMARY  ] Total 176 cases of 1 op(s).
ALL PASSED.
[==========] 176 test cases from 1 test suite ran.
[  PASSED  ] 176 test cases.

Platform：MLU590

release_test cases
[----------] Global test environment tear-down
[ SUMMARY  ] Total 50 cases of 1 op(s).
ALL PASSED.
[==========] 50 test cases from 1 test suite ran.
[  PASSED  ] 50 test cases.

release_temp cases
[----------] Global test environment tear-down
[ SUMMARY  ] Total 176 cases of 1 op(s).
ALL PASSED.
[==========] 176 test cases from 1 test suite ran.
[  PASSED  ] 176 test cases.

### 3.4 Summary Analysis

All cases are passed.

### CODE REVIEW

File: bangc-ops/mlu_op.h [hunk 1]
```diff
@@ -6984,6 +6984,129 @@ mluOpActiveRotatedFilterForward(const mluOpHandle_t handle,
                                 const mluOpTensorDescriptor_t output_desc,
                                 void *output);
 
+// Group:ThreeNNForward
+/*!
+ * @brief Returns in \b workspace_size the size of the MLU memory that is used as an extra
+ * workspace to optimize the ThreeNNForward operation. The size of the extra workspace is
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:6990]: ThreeNNForward 补充完整算子的名称，加::使用链接。

File: bangc-ops/mlu_op.h [hunk 2]
```diff
@@ -6984,6 +6984,129 @@ mluOpActiveRotatedFilterForward(const mluOpHandle_t handle,
...
+ * based on the given information of the ThreeNNForward operation, including the input
+ * tensor descriptor \b known_desc. For more information about the workspace, see
+ * "Cambricon BANGC OPS User Guide".
+ *
+ * @param[in] handle
+ * Handle to an MLUOP context that is used to manage MLU devices and queues in the
+ * ThreeNNForward operation. For detailed information, see ::mluOpHandle_t.
+ * @param[in] known_desc
+ * The escriptor of input data \b known, which contains dimension, data type and data layout.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:6999]: escriptor  > 拼写错误   descriptor

File: bangc-ops/mlu_op.h [hunk 3]
```diff
@@ -6984,6 +6984,129 @@ mluOpActiveRotatedFilterForward(const mluOpHandle_t handle,
...
+ * @param[out] workspace_size
+ * A host pointer to the returned size of the extra workspace in bytes that is used in
+ * the ThreeNNForward operation.
+ *
+ * @par Return
+ * - ::MLUOP_STATUS_SUCCESS, ::MLUOP_STATUS_BAD_PARAM
+ *
+ * @par API Dependency
+ * - This function must be called before the mluOpThreeNNForward function.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:7008]: ::mluOpThreeNNForward

File: bangc-ops/mlu_op.h [hunk 4]
```diff
@@ -6984,6 +6984,129 @@ mluOpActiveRotatedFilterForward(const mluOpHandle_t handle,
...
+ * - None.
+ */
+mluOpStatus_t MLUOP_WIN_API
+mluOpGetThreeNNForwardWorkspaceSize(const mluOpHandle_t handle,
+                                    const mluOpTensorDescriptor_t known_desc,
+                                    size_t *workspace_size);
+
+// Group:ThreeNNForward
+/*!
+ * @brief Finds closest 3 points of \b unknown among \b known, and output \b dist and index \b idx
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:7026]: closest 3 points指的是什么呢？

File: bangc-ops/mlu_op.h [hunk 5]
```diff
@@ -6984,6 +6984,129 @@ mluOpActiveRotatedFilterForward(const mluOpHandle_t handle,
...
+ * tensor. This function firstly compute dist of each known point to a unknown point, and find
+ * the closest 3 points, and output the dist and index of the known point in known dataset.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:7028]: computes      finds    outputs

File: bangc-ops/mlu_op.h [hunk 6]
```diff
@@ -6984,6 +6984,129 @@ mluOpActiveRotatedFilterForward(const mluOpHandle_t handle,
...
+ * - The shape of \b unknown, \b dist2 and \b idx should be [b, n, 3].
+ * - The shape of \b known should be [b, m, 3].
+ * - The shape of \b unknown, \b dist2, \b idx and \b known dims[0](b) should be equal.
+ * - The shape of \b unknown, \b dist2, \b idx and \b known dims[2](3) should be equal to 3.
+ * - The shape of \b unknown, \b dist2, \b idx and \b known dims[1](n) should be equal and larger
+ *   than 0.
+ *
+ * @par API Dependency
+ * - Before calling this function you need to call mluOpGetThreeNNForwardWorkspaceSize
+ *   to get the extra space size needed in mluOpThreeNNForward operation.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h]: ::mluOpThreeNNForward

File: bangc-ops/mlu_op.h [hunk 7]
```diff
@@ -6984,6 +6984,129 @@ mluOpActiveRotatedFilterForward(const mluOpHandle_t handle,
...
+ * Pointer to the MLU memory that stores the unknown tensor.
+ * @param[in] known_desc
+ * The descriptor of input data \b known, which contains dimension, data type and data layout.
+ * @param[in] known
+ * Pointer to the MLU memory that stores the known tensor.
+ * @param[in] workspace
+ * Pointer to the MLU memory that is used as an extra workspace for the ThreeNNForward
+ * operation. For more information about workspace, see "Cambricon BANGC OPS User Guide".
+ * @param[in] workspace_size
+ * The size of the extra workspace in bytes that is used in the ThreeNNForward operation.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h]: ThreeNNForward统一都修改一下

**Review by GraceGuanH — COMMENTED** [review]
Please check the comments.

File: bangc-ops/mlu_op.h (hunk 1 above)
Comment (DanieeelLiu) [reply bangc-ops/mlu_op.h]: Fixed

File: bangc-ops/mlu_op.h (hunk 2 above)
Comment (DanieeelLiu) [reply bangc-ops/mlu_op.h]: Fixed

File: bangc-ops/mlu_op.h (hunk 3 above)
Comment (DanieeelLiu) [reply bangc-ops/mlu_op.h]: Fixed

File: bangc-ops/mlu_op.h (hunk 4 above)
Comment (DanieeelLiu) [reply bangc-ops/mlu_op.h]: Fixed

File: bangc-ops/mlu_op.h (hunk 5 above)
Comment (DanieeelLiu) [reply bangc-ops/mlu_op.h]: Fixed

File: bangc-ops/mlu_op.h [hunk 8]
```diff
@@ -6984,6 +6984,129 @@ mluOpActiveRotatedFilterForward(const mluOpHandle_t handle,
...
+ * @par Scale Limitation
+ * - The shape of \b unknown, \b dist2 and \b idx should be [b, n, 3].
+ * - The shape of \b known should be [b, m, 3].
+ * - The shape of \b unknown, \b dist2, \b idx and \b known dims[0](b) should be equal.
+ * - The shape of \b unknown, \b dist2, \b idx and \b known dims[2](3) should be equal to 3.
+ * - The shape of \b unknown, \b dist2, \b idx and \b known dims[1](n) should be equal and larger
+ *   than 0.
+ *
+ * @par API Dependency
+ * - Before calling this function you need to call mluOpGetThreeNNForwardWorkspaceSize
```
[the remark this answers is not in the document]
Comment (DanieeelLiu) [reply bangc-ops/mlu_op.h]: Fixed

File: bangc-ops/mlu_op.h (hunk 6 above)
Comment (DanieeelLiu) [reply bangc-ops/mlu_op.h]: Fixed

File: bangc-ops/mlu_op.h (hunk 7 above)
Comment (DanieeelLiu) [reply bangc-ops/mlu_op.h]: Fixed

**Review by GraceGuanH — APPROVED** [review]

**Review by mahxn0 — APPROVED** [review]

**Review by WangYuan-cn — APPROVED** [review]

**Pull request merged**